app/util/appUtil.js

```diff
@@ -75,21 +91,77 @@ function formatHash(filestoreResponse) {
 }
 
 async function processMetadata(metadata, files) {
-  return Object.fromEntries(
+  const metadataItems = Object.entries(metadata)
+  if (metadataItems.length > MAX_METADATA_COUNT)
+    throw new Error(`Metadata has too many items: ${metadataItems.length}. Max item count: ${MAX_METADATA_COUNT}`)
+
+  return new Map(
     await Promise.all(
-      Object.entries(metadata).map(async ([key, filename]) => {
-        if (key.length > METADATA_KEY_LENGTH)
-          throw new Error(`Key: ${key} is too long. Maximum key length is ${METADATA_KEY_LENGTH}`)
-
-        const file = files[filename]
-        if (!file) throw new Error(`Error no attached file found for ${filename}`)
-        const filestoreResponse = await addFile(file)
-        return [key, formatHash(filestoreResponse)]
+      metadataItems.map(async ([key, value]) => {
+        const lengthInHex = new TextEncoder('hex').encode(key).length
```

Comment:
I think there's some deduplication that can be done here. You are turning the key to a `UInt8Array`lower down and this could probably just happen here for the length check

Reply:
done, changed `utf8ToUint8Array` to throw an error if too long